soccer/Referee.cpp

```diff
@@ -110,103 +29,151 @@ static const bool CancelBallPlaceOnHalt = true;
+                  << " bytes from " << _sender_endpoint << std::endl;
+        std::cerr << "Address: " << &RefereeAddress << std::endl;
+        return;
+    }
 
-    _running = true;
-    while (_running) {
-        if (!_useExternalRef) continue;
+    std::unique_lock<std::mutex> lock{_mutex};
+    _packets.push_back(packet);
+
+    stage_ = static_cast<Stage>(packet.wrapper.stage());
+    command_ = static_cast<Command>(packet.wrapper.command());
+    sent_time =
+        RJ::Time(std::chrono::microseconds(packet.wrapper.packet_timestamp()));
+    stage_time_left =
+        std::chrono::milliseconds(packet.wrapper.stage_time_left());
+    command_counter = packet.wrapper.command_counter();
+    command_timestamp =
+        RJ::Time(std::chrono::microseconds(packet.wrapper.command_timestamp()));
+    yellow_info.ParseRefboxPacket(packet.wrapper.yellow());
+    blue_info.ParseRefboxPacket(packet.wrapper.blue());
+    ballPlacementx = packet.wrapper.designated_position().x();
+    ballPlacementy = packet.wrapper.designated_position().y();
+
+    // If we have no name, we're using a default config and there's no
+    // sense trying to match the referee's output (because chances are
+    // everything is just on default configuration with no names, so
+    // more than one team/soccer instance will be trying to use the same
+    // color)
+    if (Team_Name_Lower.length() > 0) {
+        // We only want to change teams if we get something that
+        // actually matches our team name (either yellow or blue).
+        // Otherwise, keep the current color.
+        if (boost::iequals(yellow_info.name, Team_Name_Lower)) {
+            blueTeam(false);
+            _isRefControlled = true;
+        } else if (boost::iequals(blue_info.name, Team_Name_Lower)) {
+            blueTeam(true);
+            _isRefControlled = true;
+        } else {
+            _isRefControlled = false;
+        }
+    }
+}
 
-        char buf[65536];
+void Referee::setupRefereeMulticast() {
+    const auto any_address = boost::asio::ip::address_v4::any();
+    boost::asio::ip::udp::endpoint listen_endpoint{any_address,
+                                                   ProtobufRefereePort};
+
+    _asio_socket.open(listen_endpoint.protocol());
+    _asio_socket.set_option(boost::asio::ip::udp::socket::reuse_address(true));
+    try {
+        _asio_socket.bind(listen_endpoint);
+    } catch (const boost::system::system_error& e) {
+        throw std::runtime_error("Failed to bind to shared referee port");
+    }
+    // Join multicast group
+    const boost::asio::ip::address multicast_address =
+        boost::asio::ip::address::from_string(RefereeAddress);
+    _asio_socket.set_option(
+        boost::asio::ip::multicast::join_group(multicast_address));
+}
 
-        if (!socket.waitForReadyRead(500)) {
-            continue;
-        }
+bool Referee::stopRequested() {
+    return _req_stop_fut.wait_for(std::chrono::seconds(0)) ==
```

Comment:
Is there no helper for this? Polling a future without blocking seems like a pretty typical thing to do.

Reply:
nope